lib: min_heap: Introduce min-heap

sayoojkkarun opened the pull request:
Add
    1. min-heap implementation.
    2. ztest for min-heap.
    3. Documentation for min-heap.

andyross commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,183 @@
+ */
+struct min_heap_node {
+	/** Key for comparison (optional) */
+	int key;
```
Not understanding the purpose behind the key field?  It's unused by the library code, which calls the comparison function unconditionally.

Sure, lots of applications can represent sorting order with a single integer, but if they do, it should be in the app-defined struct **containing** the node, not here (where it likely just wastes space).

sayooj-aerlync replied:
Understood. I have removed it.

andyross commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,183 @@
+	/** Current elements count */
+	size_t size;
+	/** Spinlock for thread safety */
+	struct k_spinlock lock;
```
Design point:  traditionally in Zephyr, data structure libraries (c.f. slist, dlist, rbtree, even sys_heap) are unsynchronized, allowing the embedding environment to decide on how best to do that.  Lots of these things end up being used inside of more complicated locking already and don't need a private lock.

Suggest removing this.  It's not wrong to do it here, but it's definitely non-standard.  And I don't see that heapsort/priority-queues are really so special that they need their own synchronization management.

andyross commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,183 @@
+ * @return 0 on success, -EINVAL on passing invalid argument.
+ */
+int min_heap_init(struct min_heap *heap, struct min_heap_node **storage,
+		   size_t capacity, min_heap_cmp_fn cmp);
```
Given the simplicity, probably best to do this here as an inline, e.g.

```c
static inline int min_heap_int(...)
{
    *heap = MIN_HEAP_INIT(...);
}
```

andyross commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,183 @@
+ */
+#define MIN_HEAP_INIT(storage, cap, cmp_func) \
+	{ .nodes = (storage), .capacity = (cap), .size = 0, .lock = {}, \
+	.cmp = (cmp_func) }
```
size and lock are already initialized to zero per the language standard, no need to specify

andyross commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,225 @@
+	size_t parent;
+
+	if (heap == NULL || heap->size == 0) {
+		return;
```
Generally defensive coding is frowned on in Zephyr.   These should be asserts.  And likely checked somewhere up the stack closer to the entrance from app code.

sayooj-aerlync replied:
Removed all the defensive coding in the library.

andyross commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,225 @@
+{
+	if (heap == NULL || storage == NULL || cmp == NULL || capacity == 0) {
+		LOG_ERR("Invalid Parameters passed to initialise min-heap");
+		return -EINVAL;
```
Same note on defensive coding.  If it's a common/expected oerr or a tricky API it's reasonable to put an assert in place, but in general we're an RTOS and we care about code size.  Needless checking for parameter validity isn't appropriate.

sayooj-aerlync replied:
Removed all the defensive coding in the library.

andyross commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,225 @@
+	struct min_heap_node *min = heap->nodes[0];
+
+	heap->nodes[0] = heap->nodes[--heap->size];
+	heap->nodes[heap->size] = NULL;
```
Not seeing the value to the NULL here?  Nothing checks for that, all pointers below `heap->size` are non-null by construction, all others are affimatively unused.

sayooj-aerlync replied:
Changed the implementation.

andyross commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,225 @@
+	return min;
+}
+
+int min_heap_remove_by_idx(struct min_heap *heap, struct min_heap_node *node)
```
Confused here.  This isn't removing "by index", it's removing by node ID.  In fact there's no API provided where you can remove (or inspect) the "Nth" node in the array.  Should just be called "`min_heap_remove()`" no?

sayooj-aerlync replied:
Yes, correct. I have modified the function.

andyross commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,225 @@
+	return 0;
+}
+
+struct min_heap_node *min_heap_pop(struct min_heap *heap)
```
This is almost identical code to min_head_remove_by_idx() and should probably be implemented in terms of it, no?

sayooj-aerlync replied:
Yes! Using `min_head_remove()` for popping the top.

andyross commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,225 @@
+	return 0;
+}
+
+int min_heap_destroy(struct min_heap *heap)
```
This doesn't actually do anything useful?  The heap doesn't have any internal resource allocation, all this is doing is writing some fields that (by definition) the app won't/shouldn't/mustn't ever read anyway.  Suggest just removing the function.

andyross reviewed (commented):
Notes and nitpicks.  The big ones are "limit defensive coding", and "separate synchronization from data structure code".

sayooj-aerlync commented on doc/services/min_heap/index.rst:
```diff
@@ -0,0 +1,137 @@
+.. _min_heap:
+
+Minimum Heap (Min-heap)
```
Changed.

sayooj-aerlync commented on lib/min_heap/Kconfig:
```diff
@@ -0,0 +1,7 @@
+config MIN_HEAP
+	bool "Min-Heap Data Structure"
+	help
+		This option enables the Min-Heap data structure Library
```
Done.

andyross reviewed (dismissed):
Looks clean, I'm out of complaints.

FWIW: for extra credit you could look at slotting this in as an optional backend for the kernel timeout subsystem.  We occasionally field complaints that the dlist it's using is algorithmically suboptimal.  A priority queue (I'm too old to say "min-heap") is the natural choice, even though most apps have only a few timeouts active and would want the simpler/low-constant-factor list.

sayooj-aerlync commented:
> Looks clean, I'm out of complaints.
> 
> FWIW: for extra credit you could look at slotting this in as an optional backend for the kernel timeout subsystem. We occasionally field complaints that the dlist it's using is algorithmically suboptimal. A priority queue (I'm too old to say "min-heap") is the natural choice, even though most apps have only a few timeouts active and would want the simpler/low-constant-factor list.

Sure I will look into timeout subsystem as soon as this gets merged! :)

cfriedt reviewed (commented):
Sorry - didn't publish this comment when I made it originally

cfriedt commented on tests/lib/min_heap/src/test_min_heap.c:
```diff
@@ -0,0 +1,154 @@
+/** @brief Buffer to hold heap node pointers */
+static struct min_heap_node *heap_buf[MAX_ITEMS];
+/** @brief Min heap instance for testing */
+static struct min_heap heap;
```
Can this min heap be defined using `MIN_HEAP_DEFINE()`?

sayooj-aerlync replied:
Added another heap instance which is initialised with `MIN_HEAP_DEFINE`

cfriedt commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,162 @@
+ * @param cap Number of elements in the array
+ * @param cmp_func Comparator function
+ */
+#define MIN_HEAP_INIT(storage, cap, cmp_func) \
```
Is it possible to add a `_DEFINE` macro maybe along with `_INIT` that would declare a named instance?

I think that is maybe a more typical behaviour for declaration-oriented macros in Zephyr.

cfriedt commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,162 @@
+ * @note All arguments must be valid. This function does not allocate memory.
+ *
+ */
+static inline void min_heap_init(struct min_heap *heap,
```
Would it be better to make this non-inline to start out with?

I can't see there being a compile-time optimization being made here, but I might be missing something.

Some reasoning behind the `_DEFINE` vs runtime `_init()` paradigm is that `_DEFINE` is done by the linker at build time, and `_init` requires code space and CPU cycles at run time. Using the `_DEFINE` paradigm is a subtle choice that helps to keep Zephyr small.

sayooj-aerlync commented:
> Sorry - didn't publish this comment when I made it originally

No Problem. I will make the changes soon.

cfriedt commented:
> > Sorry - didn't publish this comment when I made it originally
> 
> No Problem. I will make the changes soon.

Thank you 🙏 - a few more comments too, but better to catch them now.

sayooj-aerlync commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,162 @@
+
+/**
+ * @brief min_heap
+ * @defgroup min_heap Min-heap service
+ * @ingroup os_services
```
Should it be `datastructure_apis` then?

sayooj-aerlync commented on include/zephyr/min_heap.h:
Moved.

sayooj-aerlync commented on doc/services/min_heap/index.rst:
```diff
@@ -0,0 +1,137 @@
+Example
+*******
+
+.. code-block:: c
```
@cfriedt 
should we keep the code in `samples/lib/` or `samples/subsys`? `samples/lib` does not exist currently.

cfriedt commented on tests/lib/min_heap/src/test_min_heap.c:
```diff
@@ -0,0 +1,154 @@
+		ret = min_heap_insert(&heap, &items[i].node);
+		zassert_equal(ret, 0,
+			      "min_heap insert failed %d", ret);
+	}
```
I think a key item to test is whether the "heap property" is satisfied after each iteration.

sayooj-aerlync replied:
Changed the complete test cases.

cfriedt commented on include/zephyr/sys/min_heap.h:
```diff
@@ -0,0 +1,173 @@
+ * @param cap  Capacity (number of nodes)
+ * @param cmp_func Comparison function
+ */
+#define MIN_HEAP_DEFINE(name, cap, cmp_func) \
```
This should likely include a size and align parameter, so that storage for each element may be allocated statically.

It would probably also be worth it to add a `_DEFINE_STATIC` variant that declares the min heap itself statically

cfriedt replied:
The size and alignment of each individual element or structure, which would contain the heap node data.

sayooj-aerlync replied:
Got it!

cfriedt commented on include/zephyr/sys/min_heap.h:
```diff
@@ -0,0 +1,173 @@
+ * @note All arguments must be valid. This function does not allocate memory.
+ *
+ */
+static void min_heap_init(struct min_heap *heap,
```
The body of this function needs to be moved to `min_heap.c`

```suggestion
void min_heap_init(struct min_heap *heap,
```

sayooj-aerlync replied:
Moved

cfriedt commented on tests/lib/min_heap/src/test_min_heap.c:
```diff
@@ -0,0 +1,154 @@
+ * Inserts predefined items into the heap and asserts that
+ * popping the heap returns them in ascending order of key.
+ */
+ZTEST(min_heap_suite, test_min_heap_ordering)
```
Can tests be made for each API function?

E.g. test_min_heap_init, test_min_heap_insert, test_min_heap_remove, test_min_heap_push, test_min_heap_pop, test_min_heap_find, etc?

Tests are usually executed in alphabetical order as well, so it can be helpful to order them that way in the file.

It may help to have a `test_min_heap_init_common()` function that is called into for verifying that the macro-based initializers also initialize a heap correctly.

I would also suggest reinitializing the min heap before each test in a `before()` function.

cfriedt commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,134 @@
+	return min_heap_remove(heap, 0);
+}
+
+bool min_heap_is_empty(struct min_heap *heap)
+{
+	return (heap->size == 0);
+}
```
this would be good to move to the header file

```suggestion
static inline bool min_heap_is_empty(struct min_heap *heap)
{
__ASSERT_NO_MSG(heap != NULL);

	return (heap->size == 0);
}
```

sayooj-aerlync replied:
Done.

cfriedt commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,134 @@
+	}
+}
+
+int min_heap_insert(struct min_heap *heap, struct min_heap_node *node)
```
Can this be renamed to `min_heap_push()`?

cfriedt replied:
There are a few reasons why I would insist on a "push" semantic

* it complements the "pop" semantic (and developers will naturally look for "push" as a result
* insertion is usually specified with an index (e.g. C++ standard and other languages)
* a min / max heap is effectively a priority queue, and the C++ standard uses push for that (as do other languages)
* even in the Wikipedia article mentioned, insertion is described with "push" operations

sayooj-aerlync replied:
Renamed to `min_heap_push()`

cfriedt commented on lib/min_heap/min_heap.c:
```diff
@@ -0,0 +1,134 @@
+	return 0;
+}
+
+struct min_heap_node *min_heap_remove(struct min_heap *heap, size_t id)
```
Personally, I think this would need to be complemented by a `_find()` function that accepts an eq function pointer and 'other' object, and returns a reference to the node.

Beyond that, a FOREACH iterator macro?

sayooj-aerlync replied:
A heap does not store items in strict sorted order—it only guarantees that the smallest (top) item is at index 0. The rest follow the heap invariant, not linear sorted order.
In that case what can `FOREACH` be used for?

We can add `_find()`

cfriedt replied:
> A heap does not store items in strict sorted order—it only guarantees that the smallest (top) item is at index 0. The rest follow the heap invariant, not linear sorted order.

Yep

> In that case what can `FOREACH` be used for?

It would be used for iterating over the elements - not necessarily in sorted order.

> We can add `_find()`

It's not absolutely necessary, but I'm not sure if `remove()` would be as useful without it.

Typically only the top is ever popped from a heap

https://wikipedia.org/wiki/Heap_(data_structure)

Otherwise search in heap is of course linear.

Maybe it would make sense to just remove `remove()`?

sayooj-aerlync replied:
Yes, Let's remove `remove()`

sayooj-aerlync commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,162 @@
+ *         positive value if @p a is greater than @p b,
+ *         zero if they are equal.
+ */
+typedef int (*min_heap_cmp_fn)(const struct min_heap_node *a,
+			       const struct min_heap_node *b);
```
Yes!

sayooj-aerlync replied:
Using void pointers now

sayooj-aerlync commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,162 @@
+ */
+struct min_heap {
+	/** Pointer to preallocated node pointer array */
+	struct min_heap_node **nodes;
```
The current design uses an invasive data structure (`struct min_heap_node`), which allows users to embed it within their own types. This enables efficient tracking within the Min-Heap without requiring additional indirection or memory allocation.
Switching to a `void*` interface would significantly alter the design and remove this benefit. Could you elaborate on the advantages you're aiming to achieve with the void* approach? That would help in assessing whether the design trade-offs are justified.

sayooj-aerlync commented on include/zephyr/min_heap.h:
```diff
@@ -0,0 +1,162 @@
+ * @{
+ */
+
+/**
+ * @brief Invasive node used inside user structs for min-heap support.
+ */
+struct min_heap_node {
+	/** Index within heap array */
+	size_t index;
+};
```
That's alright! I will change it!

kartben commented on doc/services/min_heap/index.rst:
```diff
@@ -0,0 +1,68 @@
+              / \     / \
+             8   9  10  12
+
+For any node at index ``i``, its children are at indices:
+- Left child:  ``2*i + 1``
```
```suggestion
For any node at index ``i``, its children are at indices:

- Left child:  ``2*i + 1``
```

kartben commented on doc/services/min_heap/index.rst:
```diff
@@ -0,0 +1,68 @@
+- Left child:  ``2*i + 1``
+- Right child: ``2*i + 2``
+
+Its parent is at index:
+- Parent: ``(i - 1) / 2``
```
also note that if you want you can use nicely formatted math formulas (https://docs.zephyrproject.org/latest/contribute/documentation/guidelines.html#mathematical-formulas)

```suggestion
Its parent is at index:

- Parent: ``(i - 1) / 2``
```

kartben reviewed (changes requested):
few doc comments - you can check CI-built version here, fwiw: https://builds.zephyrproject.io/zephyr/pr/90251/docs/services/min_heap/index.html

sayooj-aerlync commented on doc/services/index.rst:
Yes.

sayooj-aerlync commented on doc/services/min_heap/index.rst:
```diff
@@ -0,0 +1,68 @@
+.. _min_heap:
```
Changed

sayooj-aerlync commented on include/zephyr/sys/min_heap.h:
```diff
@@ -0,0 +1,173 @@
+ *         positive value if @p a is greater than @p b,
+ *         zero if they are equal.
+ */
+typedef int (*min_heap_cmp_fn)(const struct min_heap_node *a,
```
Changed!

kartben commented on doc/kernel/data_structures/min_heap.rst:
```diff
@@ -0,0 +1,77 @@
+Samples
+*******
+
+ :zephyr:code-sample: `min-heap` sample demos the API usage of Min-Heap
```
```suggestion
 :zephyr:code-sample:`min-heap` sample demos the API usage of Min-Heap
```

andyross reviewed (dismissed):
Refresh +1.  Seems like the remaining points are all doc/structure, so feel free to assume the +1 from me remains, absent non-trivial changes to the code.

cfriedt commented on include/zephyr/sys/min_heap.h:
```diff
@@ -0,0 +1,247 @@
+ *
+ * @return Pointer to the element at the given index.
+ */
+static inline void *get_element(const struct min_heap *heap, size_t index)
```
This function should be prefixed and / or perhaps in the .c file
```suggestion
static inline void *min_heap_get_element(const struct min_heap *heap, size_t index)
```

sayooj-aerlync replied:
I will prefix it, as _FOR_EACH macro is using the function to parse the heap.

sayooj-aerlync replied:
Changed.

cfriedt reviewed (changes requested):
`get_element()` should probably be prefixed and maybe put in the c file

sayooj-aerlync commented on samples/data_structures/min-heap/prj.conf:
added.

sayooj-aerlync commented on tests/lib/min_heap/prj.conf:
Added

cfriedt commented on tests/lib/min_heap/src/test_min_heap.c:
```diff
@@ -0,0 +1,131 @@
+	validate_heap_order(&my_heap);
+	zassert_true(min_heap_is_empty(&my_heap) != false, "Empty check fail");
+}
+
```
Looks really good from my perspective.

A couple of final requests 😅

Please move this test suite to tests/unit, since it's completely platform independent. That should simplify CI a bit.

Lastly, is it possible to add a second, nearly identical heap to the testsuite, with the main difference being the comparator? So, rather than e.g. `>=` as positive logic, use `<`?

sayooj-aerlync replied:
facing an issue while moving the test case to tests/unit. as for the board `unit_testing`, the macro `ARCH_STACK_PTR_ALIGN` is not defined.
Can we keep the test suite in tests/lib itself?

I am working on the positive logic.

sayooj-aerlync replied:
Changed the logic. I can take up the `ARCH_STACK_PTR_ALIGN` macro as a separate issue if it is a real one. Fix and move the code later.

sayooj-aerlync commented on tests/lib/min_heap/src/test_min_heap.c:
```diff
@@ -0,0 +1,131 @@
+MIN_HEAP_DEFINE_STATIC(my_heap, HEAP_CAPACITY, sizeof(struct data),
+		       __alignof__(struct data), compare);
+
+struct data elements[] = {
```
This is just quasi random.

nashif reviewed (changes requested):
- please split this commit, into implementation, tests, samples etc
- Please expand in the commit on what min-heap is, there are no details what that is right now.
- The kconfig should also have more information about what that is
- The name min-heaps naming along side existing heaps, the lack of details, ie.. in the kconfig might leads to confusion, although those are 2 different things, maybe move it to lib/util where we have other data structure defined there.

sayoojkkarun committed:
lib: min_heap: Add min-heap data structure

Introduce a binary min-heap implementation as a generic data structure
for use in kernel and application code. A min-heap always maintains
the smallest element at the root, making insertion and removal of the
minimum element efficient (O(log n)).

The API allows both static and dynamic initialization, supports custom
comparators.

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sayoojkkarun committed:
tests: lib: min_heap Add unit tests for min-heap API

Introduce test suite using Ztest for validating the min-heap API.
Covers insertion, removal, ordering, and custom comparators.

The tests verify correct behavior under both static and dynamic
initialization modes.

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sayoojkkarun committed:
samples: lib: min_heap: Add min-heap usage example

Introduce a simple sample application demonstrating the usage of the
min-heap API. The sample performs basic insert and remove operations
and logs the results to the console.

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sayoojkkarun committed:
doc: kernel: data_structures: Add documentation for min-heap

Includes explanation of the min-heap’s properties, constraints and
explains about the uses cases of min-heap. It also contains references
to Sample Application and min-heap API

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sayooj-aerlync commented:
> * please split this commit, into implementation, tests, samples etc
> * Please expand in the commit on what min-heap is, there are no details what that is right now.
> * The kconfig should also have more information about what that is
> * The name min-heaps naming along side existing heaps, the lack of details, ie.. in the kconfig might leads to confusion, although those are 2 different things, maybe move it to lib/util where we have other data structure defined there.

Please check. I have split the the pull request into 4 commits. 
1. Library Implementation 
2. Unit Test Application 
3. Sample Application 
4. Documentation. 
I have also added detailed Help information in Kconfig

sayooj-aerlync commented:
@andyross @nashif @cfriedt  The twister failure is unrelated to the pull request. How do we proceed on this?
`sample.kernel.bootargs.efi` on `qemu_x86_64/atom`

nashif reviewed (approved):
- please split this commit, into implementation, tests, samples etc
- Please expand in the commit on what min-heap is, there are no details what that is right now.
- The kconfig should also have more information about what that is
- The name min-heap naming is a bit confusing and might be confused with heaps, ie.. this adds min_heap into lib, where we have heaps already, the lack of details, ie.. in the kconfig might leads to confusing.

JarmouniA reviewed (approved):
Great job!

cfriedt commented:
@sayooj-aerlync - for safety-critical applications, the `while(true)` is a bit of a red flag in `heapify_down()`

Aside from moving to the `unit_testing` platform, I think in a subsequent PR, it would be good to look at merging the implementation with this, which has already been audited.

https://github.com/zephyrproject-rtos/zephyr/blob/e571f8bab4281fd03a0a0fe74def421f02451065/lib/libc/minimal/source/stdlib/qsort.c#L58

I'm on the fence about whether shifts are better or worse, since most compilers should figure that out.
